README
This library contains an R implementation of the S-Plus function interp().

I used Fortran code for cubic spline interpolation from H. Akima available 
at http://www.netlib.org/toms .
There are two version, one from 1978 (AMS code 526, used in SPlus 3.x) and a 
newer one (AMS 761) from 1996. The original versions are in the directory orig/.

I only had to write a subroutine idptli.f for linear interpolation within
triangles from the delauney triangulation.

I added  a function interpp(), which evaluates the interpolated function at 
arbitraryly choosen points and generates no regular grid as interp() does.

interp() actually calls interp.new() or interp.old() depending on its arguments,
linear interpolation is done by interp.old(), spline interpolation by interp.new
.

UPDATE (13.12.2004):

Thomas Petzoldt <petzoldt@rcs.urz.tu-dresden.de> provided an R interface to
Akimas univariate spline interpolation functions (aspline) which has been 
included into this library.

The original papers from H. Akima are available (access is not free, depends on your
ACM subscription):
TOMS 433: http://dl.acm.org/citation.cfm?id=355604.355605
TOMS 526: http://dl.acm.org/citation.cfm?id=355780.355787
TOMS 697: http://dl.acm.org/citation.cfm?id=114697.116808
TOMS 761: http://dl.acm.org/citation.cfm?id=232826.232856

Original Fortran code is freely available at Netlib/TOMS -- Collected Algorithms 
of the ACM at:
http://www.netlib.org/toms/433
http://www.netlib.org/toms/526
http://www.netlib.org/toms/697
http://www.netlib.org/toms/761

UPDATE: (28.02.2013)

Add bicubic interpolation for regular grid (TOMS 760).

UPDATE: (09.12.2016)

Only Akimas new implementation of bicubic interpolation for regular grid (TOMS 761) 
remains in the package, the old one is no longer needed. Linear interpolation is now
calcaluted based on Renkas trmesh function.

A simple bilinear interpolation has been added (no Akima code).

------------------------------------------------------------------------------
Albrecht Gebhardt          email: albrecht.gebhardt@aau.at
Institut fuer Mathematik   Tel. : (++43 463) 2700/3118
Universitaet Klagenfurt    Fax  : (++43 463) 2700/3198
Universitaetsstr. 65-67    WWW  : http://www.stat.aau.at/~agebhard
A-9020 Klagenfurt, Austria
------------------------------------------------------------------------------




Details:

See ChangeLog for details about further versions

Version 0.1-1

I used the S-Plus version of interp() (S-Plus 3.3 f. Win 3.11) as starting
point. Then I searched for Akimas idsfft function, which is referenced in 
interp().

I found the appropriate Fortran code (1978) in the ACM Collected 
Algorithms archive under

   http://www.netlib.org/toms/526

(also included here) and splitted it into the Fortran files under src/ . 
The test driver ttidbs was also included in older versions and could be 
compiled with "make test", now it is removed here.

However, it seems that this code differs a little bit from that used in S-Plus:
It implements "only" bicubic spline interpolation, no linear interpolation as
in S-Plus.

So I modified IDSFFT and added a subroutine IDPTLI which does linear 
interpolation within the triangles, generated by IDTANG.

Further changes are
+ REAL -> DOUBLE PRECISION 
+ static DIMENSIONs replaced with dynamic 
+ option to toggle extrapolation outside of convex hull added in IDSFFT and 
  IDBVIP.
  Because I don't know how to generate NAs in Fortran (I use g77 on Linux),
  I added a logical array MISSI, that indicates if a returned value should
  be NA. These values will be set to NA after the Fortran call.
+ option to handle duplicate data points added (according to an example in the 
  S-Plus help page)
+ man pages converted and rewritten
+ data set akima (from S-Plus) added 
+ function interpp() added, it evaluates the interpolated function at 
  arbitraryly choosen points and generates no regular grid as interp() does.
  There where some problems with interpp() when using the Fortran version:
   -   it crashes when compiled with "g77 -O2 -fpic" and called with more than 
       one output point.
   -   compilation with "g77 -g -fpic" fails (see src/Makefile for details)
   -   compilation with "g77 -g" works (and no crashes occur)
  These problems do not occur in the C Version (generated by f2c), so it would 
  be better to use only the src-c tree. 



After I finished the above steps I found a more recent version (ACM 761, 1996)
of Akimas interpolation code which uses the tripack package (also available at 
ACM as algorithm no. 751) for triangulation and now I'm trying to use it for
the next version of interp(). 



ChangeLog
2016-12-16  Albrecht Gebhardt  <agebhard@aau.at>

	* DESCRIPTION: fix ACM numbers

	* .Rbuildignore, NAMESPACE, R/franke.R, R/interp.R, R/interpp.R, data/franke.rda, man/franke.data.Rd, man/interp.Rd, man/interpp.Rd, src/akima433.f, src/akima697.f, src/akima761.f:
	* Fix ACM license URL
	* remove old Akima code and interface, everything is now handled by
	  Akima new code
	* Added Frankes interpolation example data (1979, see acording man page)
	* Added jitter option to interp to allow for collinear data, see examples

2016-12-10  Albrecht Gebhardt  <agebhard@aau.at>

	* LICENSE, R/bicubic.R, R/bilinear.R, R/interp.new.R, R/interp.old.R, R/interpp.new.R, R/interpp.old.R, README, man/bicubic.Rd, man/bicubic.grid.Rd, man/bilinear.Rd, man/bilinear.grid.Rd, src/akima.h, src/akima.new.f, src/akima760.f, src/bilinear.f, src/idbvip.f, src/idcldp.f, src/idgrid.f, src/idlctn.f, src/idpdrv.f, src/idptip.f, src/idptli.f, src/idsfft.f, src/idtang.f, src/idxchg.f, src/init.c, src/toms760.f:
	* Fixes for R 3.4.0 (gfortran specifc)
	* old Akima code removed
	* interp.old removed
	* interp and interp.new merged
	* similar for interpp
	* a bilinear interpolation fro rectangular grids added (also gridwise in bilinear.grid)
	* new ACM license url

2015-09-14  Albrecht Gebhardt  <agebhard@aau.at>

	* PORTING, TITLE, cleanup, src/ttidbs.f: upload

2015-08-26  Albrecht Gebhardt  <agebhard@aau.at>

	* orig/433, orig/526, orig/697, orig/761, orig/akima.bib: deleted

2013-09-16  Albrecht Gebhardt  <agebhard@aau.at>

	* acinclude.m4, configure, configure.ac, tools/acinclude.m4:
	cleanup files

2013-03-01  Albrecht Gebhardt  <agebhard@aau.at>

	* data/akima760.rda, man/akima760.Rd: add forgotten files to CVS

2013-01-20  Albrecht Gebhardt  <agebhard@aau.at>

	* R/interp2xyz.R, man/interp2xyz.Rd:
	add a convenience function to convert interp output into 3-column matrices
	suggested by M. Maechler

2012-01-07  Albrecht Gebhardt  <agebhard@aau.at>

	* src/tripack.f: second fix for 2.14.x

2012-01-06  Albrecht Gebhardt  <agebhard@aau.at>

	* R/zzz.R, data/00Index, data/akima.R, data/akima.rda:
	Changes for R 2.14.x

2010-01-14  Albrecht Gebhardt  <agebhard@aau.at>

	* src/Makevars.in:
	add autoconf stuff to detect problematic compiler options on sun studio
	compilers

2009-11-04  Albrecht Gebhardt  <agebhard@aau.at>

	* man/akima.Rd: fix documentation

2009-06-11  Albrecht Gebhardt  <agebhard@aau.at>

	* INDEX: date updated

2005-01-09  Albrecht Gebhardt  <agebhard@aau.at>

	* R/aspline.R, man/aspline.Rd: final changes before upload of akima 0.4

2001-01-10  Albrecht Gebhardt  <agebhard@aau.at>

	* src/tripack.subset.f:
	updated to use the same TRIPACK functions as package tripack.
	works also with crosslinking aklima against tripack (R-genlibs patch at
	ftp://ftp-stat.uni.klu.ac.at/pub/R/contrib). In this case src/tripack.f
	can be removed.

2000-06-04  Albrecht Gebhardt  <agebhard@aau.at>

	* data/index.doc: last changes to make "R CMD build" happy

2000-06-03  Albrecht Gebhardt  <agebhard@aau.at>

	* src-c/idptip.c, src-c/idptli.c, src-c/idsfft.c, src-c/idtang.c, src-c/idxchg.c, src-c/tripack.subset.c, src-c/ttidbs.c, src-c/Makefile, src-c/akima.new.c, src-c/driver.c, src-c/idbvip.c, src-c/idcldp.c, src-c/idgrid.c, src-c/idlctn.c, src-c/idpdrv.c:
	f2c parts removed

	* src/Makefile: New version is 0.2-4
	COMMON/IDLC/N -> COMMON/IDLC/NIT, idummy in idbvip.f
	Fix problems with \example in interp.Rd

1998-05-24  Albrecht Gebhardt  <agebhard@aau.at>

	* ChangeLog: removed from CVS

1998-03-01  Albrecht Gebhardt  <agebhard@aau.at>

	* 526: Added new spline interpolation code from Akima 1996



src/init.c

/*
 * This file contains declarations to interface R with Akimas
 * interpolation packages from ACMs Collected Algorithms Archive.
 *
 * The ACM copyright of the original Fortran code is described at
 *   http://www.acm.org/publications/policies/software-copyright-notice
 *
 * This interface layer is copyright by
 *
 *   Albrecht Gebhardt <albrecht.gebhardt@uni-klu.ac.at>
 *
 * and can be used under both GPL or ACM license to be conform
 * with the restrictions in the above mentioned ACM license.
 *
 */

#include <R.h>
#include <Rinternals.h>

#include "akima.h"

#include <R_ext/Rdynload.h>

/* Fortran interface descriptions: */

/* ACM 526, soon to be removed */
static R_NativePrimitiveArgType idbvip_t[13] = {
  INTSXP,  /* MD,   */
  INTSXP,  /* NCP,  */
  INTSXP,  /* NDP,  */
  REALSXP, /* XD,   */
  REALSXP, /* YD,   */
  REALSXP, /* ZD,   */
  INTSXP,  /* NIP,  */
  REALSXP, /* XI,   */
  REALSXP, /* YI,   */
  REALSXP, /* ZI,   */
  INTSXP,  /* IWK,  */
  REALSXP, /* WK,   */
  LGLSXP   /* MISSI */
};

static R_NativePrimitiveArgType idsfft_t[14] = {
  INTSXP,  /* MD,   */
  INTSXP,  /* NCP,  */
  INTSXP,  /* NDP,  */
  REALSXP, /* XD,   */
  REALSXP, /* YD,   */
  REALSXP, /* ZD,   */
  INTSXP,  /* NXI,  */
  INTSXP,  /* NYI,  */
  REALSXP, /* XI,   */
  REALSXP, /* YI,   */
  REALSXP, /* ZI,   */
  INTSXP,  /* IWK,  */
  REALSXP, /* WK,   */
  LGLSXP   /* MISSI */
};

/* ACM 760: */
static R_NativePrimitiveArgType rgbi3p_t[12] = {
  INTSXP,  /* MD,     */
  INTSXP,  /* NXP,    */
  INTSXP,  /* NYP,    */
  REALSXP, /* XD,     */
  REALSXP, /* YD,     */
  REALSXP, /* ZD,     */
  INTSXP,  /* NIP,    */
  REALSXP, /* XI,     */
  REALSXP, /* YI,     */
  REALSXP, /* ZI,     */
  INTSXP,  /* IER,    */
  REALSXP  /* WK,     */
};

/* ACM 761: */
static R_NativePrimitiveArgType sdbi3p_t[17] = {
  INTSXP,  /* MD,     */
  INTSXP,  /* NDP,    */
  REALSXP, /* XD,     */
  REALSXP, /* YD,     */
  REALSXP, /* ZD,     */
  INTSXP,  /* NIP,    */
  REALSXP, /* XI,     */
  REALSXP, /* YI,     */
  REALSXP, /* ZI,     */
  INTSXP,  /* IER,    */
  REALSXP, /* WK,     */
  INTSXP,  /* IWK,    */
  LGLSXP,  /* EXTRPI, */
  INTSXP,  /* NEAR,   */
  INTSXP,  /* NEXT,   */
  REALSXP, /* DIST    */
  LGLSXP   /* LINEAR  */
};

static R_NativePrimitiveArgType sdsf3p_t[18] = {
  INTSXP,  /* MD,     */
  INTSXP,  /* NDP,    */
  REALSXP, /* XD,     */
  REALSXP, /* YD,     */
  REALSXP, /* ZD,     */
  INTSXP,  /* NXI,    */
  REALSXP, /* XI,     */
  INTSXP,  /* NYI,    */
  REALSXP, /* YI,     */
  REALSXP, /* ZI,     */
  INTSXP,  /* IER,    */
  REALSXP, /* WK,     */
  INTSXP,  /* IWK,    */
  LGLSXP,  /* EXTRPI, */
  INTSXP,  /* NEAR,   */
  INTSXP,  /* NEXT,   */
  REALSXP, /* DIST    */
  LGLSXP   /* LINEAR  */
};

/* ACM 697: */
static R_NativePrimitiveArgType uvip3p_t[8] = {
  INTSXP,  /* NP, */
  INTSXP,  /* ND, */
  REALSXP, /* XD, */
  REALSXP, /* YD, */
  INTSXP,  /* NI, */
  REALSXP, /* XI, */
  REALSXP, /* YI, */
  INTSXP   /* ERR */
};

/* ACM 433: */
static R_NativePrimitiveArgType intrpl_t[7] = {
  INTSXP,  /* L,  */
  REALSXP, /* X,  */
  REALSXP, /* Y,  */
  INTSXP,  /* N,  */
  REALSXP, /* U,  */
  REALSXP, /* V,  */
  INTSXP   /* ERR */
};

/* A. Gebhardt, no ACM code: */
static R_NativePrimitiveArgType biliip_t[9] = {
  REALSXP, /* X0,  */
  REALSXP, /* Y0,  */
  REALSXP, /* Z0,  */
  INTSXP,  /* N0,  */
  REALSXP, /* X,   */
  REALSXP, /* Y,   */
  REALSXP, /* Z,   */
  INTSXP,  /* NX,  */
  INTSXP   /* NY   */
};


static R_FortranMethodDef fortranMethods[] = {
  {"idbvip", (DL_FUNC) &F77_SUB(idbvip), 13, idbvip_t}, /* interpp.old, scheduled for removal */
  {"idsfft", (DL_FUNC) &F77_SUB(idsfft), 14, idsfft_t}, /* interp.old, scheduled for removal  */
  {"sdbi3p", (DL_FUNC) &F77_SUB(sdbi3p), 17, sdbi3p_t}, /* interpp     */
  {"sdsf3p", (DL_FUNC) &F77_SUB(sdsf3p), 18, sdsf3p_t}, /* interp      */
  {"uvip3p", (DL_FUNC) &F77_SUB(uvip3p), 8,  uvip3p_t}, /* aspline     */
  {"intrpl", (DL_FUNC) &F77_SUB(intrpl), 7,  intrpl_t}, /* aspline     */
  {"rgbi3p", (DL_FUNC) &F77_SUB(rgbi3p), 12, rgbi3p_t}, /* bicubic     */
  {"biliip", (DL_FUNC) &F77_SUB(biliip), 9,  biliip_t}, /* bilinear    */
  {NULL, NULL, 0}
};

void
R_init_akima(DllInfo *info)
{
  R_registerRoutines(info,
		     NULL /*cMethods*/, NULL /*callMethods*/,
		     fortranMethods, NULL/*externalMethods*/);
}
